In [185]:
import urllib.request
import math as mt
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [186]:
def mean(arr):
    sum = 0
    for el in arr:
        sum += el
    return sum/len(arr)

def var(arr,sample = False):
    var_sum = 0
    mu = mean(arr)
    for el in arr:
        var_sum += (el - mu) ** 2
    divisor = len(arr)
    if sample:
        divisor -= 1 #不偏
    return var_sum/divisor

def sd(arr, sample = False):
    return mt.sqrt(var(arr, sample))

def  stanerr(arr):
    return sd(arr, True)/mt.sqrt(len(arr))

def sos(arr):
    return var(arr) * len(arr)

In [187]:
#課題6.1
wakuwaku = [80, 75, 80, 90, 95, 80, 80, 85, 85, 80, 90, 80, 75, 90, 85, 85, 90, 90, 85, 80]
mogumogu = [75, 70, 80, 85, 90, 75, 85, 80, 80, 75, 80, 75, 70, 85, 80, 75, 80, 80, 90, 80]
pakupaku = [80, 80, 80, 90, 95, 85, 95, 90, 85, 90, 95, 85, 98, 95, 85, 85, 90, 90, 85, 85]

In [188]:
mean(wakuwaku), mean(mogumogu), mean(pakupaku)

(84.0, 79.5, 88.15)

In [189]:
sd(wakuwaku), sd(mogumogu), sd(pakupaku)

(5.385164807134504, 5.454356057317857, 5.341114115987412)

In [190]:
mean(wakuwaku + mogumogu + pakupaku), sd(wakuwaku + mogumogu + pakupaku)

(83.88333333333334, 6.447458586313078)

In [191]:
# 対立仮説
# 少なくとも、3つのお店のポテトの評価（母集団）の平均の一つの組み合わせにおいて差はある

In [192]:
#課題6.2
wakuwaku = [80, 75, 80, 90, 95, 80, 80, 85, 85, 80, 90, 80, 75, 90, 85, 85, 90, 90, 85, 80]
mogumogu = [75, 70, 80, 85, 90, 75, 85, 80, 80, 75, 80, 75, 70, 85, 80, 75, 80, 80, 90, 80]
pakupaku = [80, 80, 80, 90, 95, 85, 95, 90, 85, 90, 95, 85, 98, 95, 85, 85, 90, 90, 85, 85]

In [193]:
# それぞれの店のポテトの点数の平均（上表 a, b, c ）

In [194]:
mean(wakuwaku), mean(mogumogu), mean(pakupaku)

(84.0, 79.5, 88.15)

In [195]:
# 全体の平均（上表 d ）

In [196]:
mean(wakuwaku + mogumogu + pakupaku)
all = wakuwaku + mogumogu + pakupaku

In [197]:
mean(all)

83.88333333333334

In [198]:
# それぞれの店のポテトの点数の偏差平方和（上表 e, f, g ）
# 全体の偏差平方和（上表 h ）

In [199]:
def deviation_sos(arr, av=None):
    if av==None:
        av = mean(arr)
    sum = 0
    for x in arr:
        sum += (av - x) ** 2
    return sum

In [200]:
def deviation_sos_cross(arr):
    all_arr = []
    sum = 0
    for arrt in arr:
        all_arr += arrt
    all_av = mean(all_arr)
    for arrt in arr:
        av = mean(arrt)
        sum += (av - all_av) ** 2 * len(arrt)
    return sum

In [201]:
deviation_sos(wakuwaku), deviation_sos(mogumogu), deviation_sos(pakupaku), deviation_sos(all)

(580.0, 595.0, 570.5499999999998, 2494.1833333333343)

In [202]:
# 群内の偏差平方和（上表　i ）

In [203]:
deviation_sos(wakuwaku) + deviation_sos(mogumogu) + deviation_sos(pakupaku)

1745.5499999999997

In [204]:
# 群間の偏差平方和（上表 j ）

In [205]:
deviation_sos_cross([wakuwaku, mogumogu, pakupaku])

748.6333333333343

自由度を求めてください。
- ( k ) 群間の自由度　=　群の数　－　1
- ( l ) 群内の自由度　=　（群1のデータ数－1）＋（群2のデータ数－1）+（群3のデータ数－1）
- ( m ) 全体の自由度　=　データの総個数　－　1

In [206]:
3 - 1

2

In [207]:
len(wakuwaku) - 1 + len(mogumogu) -1 + len(pakupaku) - 1

57

In [208]:
len(all) - 1

59

 群間の平均平方

In [209]:
a = deviation_sos_cross([wakuwaku, mogumogu, pakupaku])/2
a

374.3166666666672

群内の平均平方

In [210]:
b = (deviation_sos(wakuwaku) + deviation_sos(mogumogu) + deviation_sos(pakupaku))/57
b

30.62368421052631

F値

In [211]:
a/b

12.223110194494591

In [212]:
# 計算したF値と、F分布表を比較
# 群間自由度2、群内自由度57、4.98
#帰無仮説を棄却して、少なくとも、3つのお店のポテトの評価の平均の一つの組み合わせにおいて差はあるということになります。

In [213]:
#課題6.4
wakuwaku = [90, 70, 75, 90, 65, 70, 80, 85, 75, 75, 85, 75, 80, 90, 80]
mogumogu = [75, 95, 75, 80, 75, 85, 75, 85, 80, 65, 90, 75, 90, 75, 85]
pakupaku = [65, 60, 75, 80, 65, 60, 70, 85, 70, 75, 75, 70, 80, 75, 80]

In [214]:
# 帰無仮説 : 3つのお店のポテトの評価の平均のどんな組み合わせにおいても差はない
# 対立仮説 : 少なくとも、3つのお店のポテトの評価の平均の一つの組み合わせにおいて差はある

In [215]:
all = wakuwaku + mogumogu + pakupaku
mean(wakuwaku), mean(mogumogu), mean(pakupaku), mean(all)

(79.0, 80.33333333333333, 72.33333333333333, 77.22222222222223)

In [216]:
var(wakuwaku), var(mogumogu), var(pakupaku), var(all)

(57.333333333333336, 58.222222222222214, 52.888888888888886, 68.39506172839505)

 <TABLE BORDER="1" align="center" CELLPADDING="2" CELLSPACING="0">
   <TR> 
     <TD WIDTH=100 BGCOLOR="#CCCCCC"> <CENTER>
         要因 </CENTER></TD>
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         偏差平方和 </CENTER></TD>
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         自由度 </CENTER></TD>
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         平均平方 </CENTER></TD>
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         F
 </CENTER></TD>
   </TR>
   <TR> 
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         群間
 </CENTER></TD>
     <TD WIDTH=100> <CENTER>551.111
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>2
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>275.56
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>4.580
         　 </CENTER></TD>
   </TR>
   <TR> 
     <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
         群内 
       </CENTER></TD>
     <TD WIDTH=100> <CENTER>2526.67
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>42
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>60.159
         　 </CENTER></TD>
   </TR>
   <TR> 
     <TD WIDTH=100 BGCOLOR="#CCCCCC"> <CENTER>
         全体 </CENTER></TD>
     <TD WIDTH=100> <CENTER>3077.78
         　 </CENTER></TD>
     <TD WIDTH=100> <CENTER>44
         　 </CENTER></TD>
   </TR>
 </TABLE>

In [217]:
#i
deviation_sos(wakuwaku) + deviation_sos(mogumogu) + deviation_sos(pakupaku)

2526.6666666666665

In [218]:
#j
deviation_sos_cross([wakuwaku,mogumogu,pakupaku])

551.1111111111114

In [219]:
#h
deviation_sos(all)

3077.7777777777774

In [220]:
#k,l,m
3-1, len(wakuwaku) - 1 + len(mogumogu) - 1 + len(pakupaku) - 1, len(all) - 1

(2, 42, 44)

In [221]:
#n,o
a,b = deviation_sos_cross([wakuwaku, mogumogu, pakupaku])/2,(deviation_sos(wakuwaku) 
                                                             + deviation_sos(mogumogu) + deviation_sos(pakupaku))/42
a,b

(275.5555555555557, 60.15873015873016)

In [222]:
#p
f = a/b
f

4.580474934036942

In [223]:
#3.23, 5.18 5%,1%

- 5%有意水準で棄却域に入って帰無仮説を棄却します。「少なくとも、3つのお店のポテトの評価（母集団）の平均の一つの組み合わせにおいて差はある」ということになります。
- 1%有意水準で棄却域に入ってないので、帰無仮説を賛成し、「3つのお店のポテトの評価の平均のどんな組み合わせにおいても差はない」ということになります。


# 第7章

In [224]:
a1b1 = [65, 85, 75, 85, 75, 80, 90, 75, 85, 65, 75, 85, 80, 85, 90] # クリスピーで辛口
a1b2 = [65, 70, 80, 75, 70, 60, 65, 70, 85, 60, 65, 75, 70, 80, 75] # クリスピーで普通味
a2b1 = [70, 65, 85, 80, 75, 65, 75, 60, 85, 65, 75, 70, 65, 80, 75] # 普通の衣で辛口
a2b2 = [70, 70, 85, 80, 65, 75, 65, 85, 80, 60, 70, 75, 70, 80, 85] # 普通の衣で普通味


  <TABLE HEIGHT="345" BORDER="1" CELLSPACING="0" 
   CELLPADDING="0">
     <TR> 
       <td bgcolor="#cccccc"></td>
       <TD BGCOLOR="#dddddd" HEIGHT="20"> <CENTER>
           クリスピーで辛口 </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           クリスピーで普通味 
         </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           普通の衣で辛口 </CENTER></TD>
       <TD WIDTH="20%" BGCOLOR="#dddddd"> <CENTER>
           普通の衣で普通味 
         </CENTER></TD>
       <TD BGCOLOR="#dddddd" HEIGHT="20"> <CENTER>
           クリスピー </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           普通の衣 
         </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           辛口 </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           普通味 
         </CENTER></TD>
       <TD BGCOLOR="#dddddd"> <CENTER>
           全体 
         </CENTER></TD>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">1</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">2</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"??> 70</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">3</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">4</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">5</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">6</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
       <TD ALIGN="CENTER"> 60</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">7</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">8</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 60</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">9</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">10</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
       <TD ALIGN="CENTER"> 60</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 60</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">11</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">12</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">13</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <TD ALIGN="CENTER"> 65</TD>
       <TD ALIGN="CENTER"> 70</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">14</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <TD ALIGN="CENTER"> 80</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <TR> 
       <td bgcolor="#cccccc">15</td>
       <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 75</TD>
       <TD ALIGN="CENTER"> 85</TD>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
       <td bgcolor="#cccccc"></td>
     </TR>
     <tr><td bgcolor="#cccccc">平均</td><td>( a )</td><td>( b )</td><td>( c )</td><td>( d )</td>
         <td>( e )</td><td>( f )</td><td>( g )</td><td> ( h )</td><td>( i )</td>
     </tr>
     <tr><td bgcolor="#cccccc">偏差平方和</td><td>( j )</td><td>( k )</td><td>( l )</td><td>( m )</td>
         <td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td>( n )</td>
     </tr>
     </TABLE>

In [225]:
# 以下の４つのうち、上２つを「要因A（食感）」、下２つを「要因B（味）」と呼びます。
a1 = a1b1 + a1b2# クリスピーのデータを作ってください。
a2 = a2b1 + a2b2# 普通の衣のデータを作ってください。
b1 = a1b1 + a2b1# 辛口のデータを作ってください。
b2 = a1b2 + a2b2# 普通味のデータを作ってください。

In [226]:
all = a1 + a2

In [227]:
# 各項目の平均値を計算してください（上表 a, b, c, d, e, f, g, h, i ）
a, b, c, d, e, f, g, h, i=mean(a1b1),mean(a1b2),mean(a2b1),mean(a2b2),\
    mean(a1),mean(a2),mean(b1),mean(b2),mean(all)
a, b, c, d, e, f, g, h, i

(79.66666666666667,
 71.0,
 72.66666666666667,
 74.33333333333333,
 75.33333333333333,
 73.5,
 76.16666666666667,
 72.66666666666667,
 74.41666666666667)

In [228]:
# 各群の偏差平方和を計算してください（上表 j, k, l, m ）
j,k,l,m = deviation_sos(a1b1),deviation_sos(a1b2),deviation_sos(a2b1),deviation_sos(a2b2) 
j,k,l,m

(873.3333333333333, 760.0, 843.3333333333333, 893.3333333333333)

In [229]:
# 全体の偏差平方和を計算してください（上表 n ）
n = deviation_sos(all)
n

4004.5833333333326

In [230]:
# 要因Aによる残差（偏差平方和）を計算してください （下表 o )

要因Aによる残差（偏差平方和）を計算してください （下表 o )

- 要因A（食感）による残差 ＝ (クリスピーの平均 ー 全体の平均)の２乗 ✕ クリスピーのデータ個数　＋　(普通の衣の平均 ー 全体の平均)の２乗　✕ 普通衣のデータ個数

In [231]:
def fact_diff(arr,arri,all=None):
    if all == None:
        all = arr + arri
    tmp1 = (mean(arr) - mean(all)) ** 2 * len(arr)
    tmp2 = (mean(arri) - mean(all)) ** 2 * len(arri)
    return tmp1 + tmp2

In [232]:
o = fact_diff(a1,a2)
o

50.4166666666664

In [233]:
# 要因Bによる残差（偏差平方和）を計算してください (下表 p )
p = fact_diff(b1,b2)
p

183.75

In [234]:
#arr = {[mure],[a1,a2],[b1,b2]}
def all_diff(arr):
    all = arr[1][0] + arr[1][1]
    all_mu = mean(all)
    sum = 0
    for x in arr[0]:
        sum += (mean(x) - all_mu) ** 2 * len(x)
    sum -= fact_diff(arr[1][0], arr[1][1]) + fact_diff(arr[2][0],arr[2][1])
    return sum

In [235]:
# 交互作用による残差（偏差平方和）を計算してください (下表 q )
q = all_diff([[a1b1,a1b2,a2b1,a2b2], [a1,a2],[b1,b2]])
q

400.41666666666754

In [236]:
s = 2 - 1
t = 2 - 1
u = s * t
v = len(all) - 1
w = v -s -t -u
print(s,t,u,w,v)

1 1 1 56 59


In [237]:
# 残りの残差（偏差平方和）を計算してください (下表 r )
r = sum([deviation_sos(a1b1),deviation_sos(a1b2)\
         ,deviation_sos(a2b1),deviation_sos(a2b2)])
print(r)

3370.0


In [238]:
# 要因A、要因B、交互作用、残差の平均平方をそれぞれ求めてください。(上表 x, y, z, aa )
x = o/s
y = p/t
z = q/u
aa = r/w
x,y,z,aa

(50.4166666666664, 183.75, 400.41666666666754, 60.17857142857143)

In [239]:
# F値を求めてください。
bb = x/aa
cc = y/aa
dd = z/aa
bb,cc,dd

(0.8377843719089966, 3.0534124629080117, 6.653808110781418)

In [240]:
# 以下の４つをそれぞれ「群」と呼びます。
a1b1 = [80, 85, 75, 70, 75, 60, 70, 75, 85, 70, 75, 85, 75, 85, 90] # ピリ辛バーガー + ピリ辛ポテト
a1b2 = [85, 65, 80, 75, 85, 65, 75, 60, 80, 70, 75, 80, 65, 85, 75] # ピリ辛バーガー + 普通のポテト
a2b1 = [60, 70, 80, 75, 65, 60, 70, 75, 80, 65, 65, 75, 70, 85, 60] # 普通のハンバーガー + ピリ辛ポテト
a2b2 = [55, 65, 50, 70, 50, 65, 75, 65, 75, 65, 70, 55, 65, 70, 65] # 普通のハンバーガー + 普通のポテト

In [241]:
# 以下の４つのうち、上２つを「要因１（バーガーの味）」、下２つを「要因２（ポテトの味）」と呼びます。
a1 = a1b1 + a1b2# ピリ辛バーガーのデータを作ってください。
a2 = a2b1 + a2b2# 普通のバーガーのデータを作ってください。
b1 = a1b1 + a2b1# ピリ辛ポテトのデータを作ってください。
b2 = a1b2 + a2b2# 普通のポテトのデータを作ってください。

In [242]:
# 各項目の平均値を計算してください（上表 a, b, c, d, e, f, g, h, i ）
a, b, c, d, e, f, g, h, i=mean(a1b1),mean(a1b2),mean(a2b1),mean(a2b2),\
    mean(a1),mean(a2),mean(b1),mean(b2),mean(all)
a, b, c, d, e, f, g, h, i

(77.0,
 74.66666666666667,
 70.33333333333333,
 64.0,
 75.83333333333333,
 67.16666666666667,
 73.66666666666667,
 69.33333333333333,
 74.41666666666667)

In [243]:
# 全体の偏差平方和を計算してください（上表 n ）
n = deviation_sos(all)
n

4004.5833333333326

In [244]:
o = fact_diff(a1,a2)
o

1126.6666666666642

In [245]:
# 要因Bによる残差（偏差平方和）を計算してください (下表 p )
p = fact_diff(b1,b2)
p

281.6666666666679

In [246]:
# 交互作用による残差（偏差平方和）を計算してください (下表 q )
q = all_diff([[a1b1,a1b2,a2b1,a2b2], [a1,a2],[b1,b2]])
q

60.00000000000182

In [247]:
s = 2 - 1
t = 2 - 1
u = s * t
v = len(all) - 1
w = v -s -t -u
print(s,t,u,w,v)

1 1 1 56 59


In [248]:
# 残りの残差（偏差平方和）を計算してください (下表 r )
r = sum([deviation_sos(a1b1),deviation_sos(a1b2)\
         ,deviation_sos(a2b1),deviation_sos(a2b2)])
print(r)

3596.666666666667


In [249]:
# 要因A、要因B、交互作用、残差の平均平方をそれぞれ求めてください。(上表 x, y, z, aa )
x = o/s
y = p/t
z = q/u
aa = r/w
x,y,z,aa

(1126.6666666666642, 281.6666666666679, 60.00000000000182, 64.22619047619048)

In [250]:
# F値を求めてください。要因A、要因B、交互作用
bb = x/aa
cc = y/aa
dd = z/aa
bb,cc,dd

(17.542168674698758, 4.385542168674717, 0.9341983317887215)

5%有意水準
- 要因A（食感）により、棄却域に入って、点数の差がある。
- 要因B（味）により、棄却域に入って、点数の差がある。
- 交互要因により、棄却域に入ってないので、、点数の差がない。<br>

1%有意水準
- 要因A（食感）により、棄却域に入って、点数の差がある。
- 要因B（味）により、棄却域に入ってないので、点数の差がない。
- 交互要因により、棄却域に入ってないので、、点数の差がない。

<table border="1" align="center" cellpadding="2" cellspacing="0">
  <tbody><tr> 
    <td width="100" rowspan="2" bgcolor="#CCCCCC"> <center>
        群内の自由度 
      </center>
      <div align="center"></div></td>
    <td colspan="5" bgcolor="#dddddd"> <center>
        群間の自由度 
      </center>
      <center>
      </center>
      <center>
      </center>
      <center>
      </center></td>
  </tr>
  <tr> 
    <td width="100" bgcolor="#dddddd"> <div align="center">1</div></td>
    <td width="100" bgcolor="#dddddd"> <div align="center">2</div></td>
    <td width="100" bgcolor="#dddddd"> <div align="center">3</div></td>
    <td width="100" bgcolor="#dddddd"> <div align="center">4</div></td>
    <td width="100" bgcolor="#dddddd"> <div align="center">5</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">10</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.96</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.10</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.71</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.48</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.33</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">20</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.35</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.49</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.10</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.87</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.71</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">30</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.17</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.32</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.92</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.69</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.53</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">40</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.08</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.23</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.84</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.61</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.45</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">50</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.03</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.18</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.79</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.56</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.40</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">60</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.00</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.15</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.76</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.53</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.37</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">70</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.98</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.13</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.74</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.50</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.35</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">80</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.96</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.11</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.72</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.49</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.33</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">90</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.95</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.10</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.71</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.47</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.32</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">100</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.94</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.09</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.70</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.46</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.31</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">200</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.89</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.04</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.65</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.42</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.26</div></td>
  </tr>
  <tr> 
    <td bgcolor="#CCCCCC"><div align="center">300</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.87</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.03</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.63</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.40</div></td>
    <td bgcolor="#eeeeee"><div align="right">2.24</div></td>
  </tr>
</tbody></table>

<table border="1" align="center" cellpadding="2" cellspacing="0">
  <tbody><tr> 
    <td width="100" rowspan="2" bgcolor="#bCCCCC"> 
      <center>
        群内の自由度 
      </center>
      <div align="center"></div></td>
    <td colspan="5" bgcolor="#bddddd"> 
      <center>
        群間の自由度 
      </center>
      <center>
      </center>
      <center>
      </center>
      <center>
      </center></td>
  </tr>
  <tr> 
    <td width="100" bgcolor="#bddddd"> 
      <div align="center">1</div></td>
    <td width="100" bgcolor="#bddddd"> 
      <div align="center">2</div></td>
    <td width="100" bgcolor="#bddddd"> 
      <div align="center">3</div></td>
    <td width="100" bgcolor="#bddddd"> 
      <div align="center">4</div></td>
    <td width="100" bgcolor="#bddddd"> 
      <div align="center">5</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">10</div></td>
    <td bgcolor="#eeeeee"><div align="right">10.04</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.56</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.55</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.99</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.64</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">20</div></td>
    <td bgcolor="#eeeeee"><div align="right">8.10</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.85</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.94</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.43</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.10</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">30</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.56</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.39</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.51</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.02</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.70</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">40</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.31</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.18</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.31</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.83</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.51</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">50</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.17</div></td>
    <td bgcolor="#eeeeee"><div align="right">5.06</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.20</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.72</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.41</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">60</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.08</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.98</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.13</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.65</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.34</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">70</div></td>
    <td bgcolor="#eeeeee"><div align="right">7.01</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.92</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.07</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.60</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.29</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">80</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.96</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.88</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.04</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.56</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.26</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">90</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.93</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.85</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.01</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.53</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.23</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">100</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.90</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.82</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.98</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.51</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.21</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">200</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.76</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.71</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.88</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.41</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.11</div></td>
  </tr>
  <tr> 
    <td bgcolor="#bCCCCC"> 
      <div align="center">300</div></td>
    <td bgcolor="#eeeeee"><div align="right">6.72</div></td>
    <td bgcolor="#eeeeee"><div align="right">4.68</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.85</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.38</div></td>
    <td bgcolor="#eeeeee"><div align="right">3.08</div></td>
  </tr>
</tbody></table>